[limits] Integrate section [limits.numeric] into section [numeric.lim…

jensmaurer opened the pull request:
…its].

Fixes #785.

tkoeppe commented:
This would be a nice cleanup. @jwakely, @zygoloid, any objections to the loss of a stable name?

jwakely commented:
Fine by me.

W-E-Brown commented:
This cleanup seems to lose the introductory nature of the first paragraph.  Perhaps let's promote the original [limits.numeric]/1 to [limits]/1 instead of moving it so far down?

> On Nov 30, 2016, at 12:53 PM, Jens Maurer <notifications@github.com> wrote:
> 
> …its].
> 
> Fixes #785.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/cplusplus/draft/pull/1159
> 
> Commit Summary
> 
> 	• [limits] Integrate section [limits.numeric] into section [numeric.limits].
> File Changes
> 
> 	• M source/support.tex (68)
> Patch Links:
> 
> 	• https://github.com/cplusplus/draft/pull/1159.patch
> 	• https://github.com/cplusplus/draft/pull/1159.diff

jensmaurer commented:
@W-E-Brown: Hm.. But the introductory sentence in the first paragraph only mentions numeric_limits, but the synopsis also has float_round_style and float_denorm_style. At this point, it's not clear these are related to the numeric_limits template. That said, I do admit there is little besides numeric_limits in the [limits] section.

tkoeppe commented:
The entire structure of 18.3.2 is weird. Anywhere else in the Standard we would structure it like this:

* 18.3.2: Numeric something something
  * 18.3.2.1: Header `<numeric>` synopsis
  * 18.3.2.2: Class template `numeric_limits`
    * Class template definition
    * 18.3.2.2.1: `numeric_limits` members
    * 18.3.2.2.2: `numeric_limits` specializations
  * 18.3.2.3: Type `float_round_style`
  * 18.3.2.4: Type `float_denorm_style`

That is, the synopsis is the first sibling in a subclause that describes the things that are listed in the synopsis. That's how we do it for most other library components. A "General" section could optionally precede the synopsis, but only if there's something useful to say. (I think the origin of "General" sections lies in the ISO drafting directives which say never to have "mixed" content in a subclause, i.e. paragraphs *and* sub-subclauses.)

tkoeppe commented:
In a hypothetical world, and given that we now have full synopses for the C library, I could even imagine the following structure:

* 18.3: Implementation properties
  * 18.3.1: General
  * 18.3.2: Header `<limits>` synopsis
  * 18.3.3: Header `<climits>` synopsis
  * 18.3.4: Header `<cfloat>` synopsis
  * 18.3.5: Class template `numeric_limits`
    * 18.3.5.1: members [etc]
  * 18.3.5: Floating point rounding and extreme values
  * 18.3.6: C library

jensmaurer commented:
A mild issue with all the presentation options so far is that float_round_style and float_denorm_style are used in the definition of numeric_limits. It would be good to introduce/explain them _before_ the numeric_limits template. Also, when extracting the <cXXX> synopses, 18.3.6 "C library" becomes empty.

I'd also like to point out that we have a subsection "C library" in clause 27, too, and we just have the synopses under this subsection with a one-sentence "same as in C" statement. Ok, but elsewhere (e.g. in the "strings" clause), the  C library stuff is just adjoint to the C++ stuff, without particularly highlighting it's from C.

jensmaurer commented:
Ok, I've done something along @tkoeppe 's "hypothetical world". I like it now. (Note that I expressly only moved sections, added/removed subsection headings, and fixed cross-references.)

W-E-Brown commented:
No longer hypothetical, eh?  I like it.

jensmaurer commented:
Rebased.

jensmaurer commented on source/support.tex:
```diff
@@ -444,18 +498,43 @@
+\tcode{numeric_limits}
+template, specializations shall define these values in such a way
+that they are usable as
+constant expressions.
```
Well, there are member functions declared "static constexpr" (such as `infinity()`), and users can specialize numeric_limits for their own types. Since a function can be declared "constexpr", yet not be usable in a constant expression, my guess is that this essentially requires that "T" doesn't access global or volatile variables when copied. (Such as copy is required for the return value of infinity(), for example.)

zygoloid replied:
Ah, this applies to user specializations. Seems weird to put a "shall" on that, but OK.

jensmaurer replied:
See also [LWG 2833](https://cplusplus.github.io/LWG/lwg-active.html#2833) "Library needs to specify what it means when it declares a function constexpr".

tkoeppe commented:
Tip for the future: Please always check the overfull box status for your proposed change:

    fgrep -e Overfull -e .tex std.log

jensmaurer commented:
Alerted Marshall Clow (LWG chair) to this change.
